### PR TITLE
ADBDEV-6599: Make gprestore --resize-cluster use --jobs for parallel restoration

### PR DESCRIPTION
**Pull request by RekGRpth**
Make gprestore --resize-cluster use --jobs for parallel restoration

Users consider the current behavior to be incorrect, since the --jobs parameter
is used when creating a backup, which is not actually used during restoration.
Restoration works in a single instance of the gpbackup_helper agent (in the
--restore-agent mode).

In short, when restoring a backup using the gprestore utility with the
--resize-cluster option, which was previously made in parallel creation mode
(--jobs), the utility does not use the corresponding mode during restoration.
In this case, restoration occurs through gpbackup_helper, which is launched in
one instance of this process for each segment. Thus, one table is processed at
a time based on the list of table OIDs passed when gpbackup_helper is launched.

As a solution to the problem, this patch implements the launch of several
gpbackup_helper instances based on the value passed in the --jobs argument.

To do this:

1) The general list of table OIDs for restoration is distributed between
several instances of gpbackup_helper processes according to the specified value
of the --jobs parameter. Also this list is distributed between goroutines in
main process by same way.

2) Since the general list of table OIDs for restoration is sorted by size, the
distribution among gpbackup_helper instances and among goroutines is done
according to the principle of sequentially distributing one table OID to each
instance and goroutine. Thus, the load is expected to be more uniform. All
batches corresponding to the same table OID go to the same gpbackup_helper
instance and same goroutine.

3) An integer identifier of the "instance number" is added to the file name of
the list of table OIDs for restoration by gpbackup_helper instances. They are
not logically related in any way, the main thing is that its own list is passed
as an argument when starting gpbackup_helper.

4) The instance number is also added to the name of the script file
(scriptFile).

5) The instance number is also added to the name of the pipeFile file, error
file and skip file.

6) To create the initial list of pipes, one pipe is created for the first table
OID (according to the sorting of table OIDs by decreasing table size), which
will be processed by each instance of gpbackup_helper. Thus, at the time of
starting each instance of gpbackup_helper, the file system will already have
the first pipe corresponding to the first table OID for this instance.

7) When starting gpbackup_helper, the value of the --jobs parameter is taken
into account (in fact, the number of connections in the pool). The number of
launched instances corresponds to the value of the parameter. The value of the
path of the first pipe (pipeFile), the path to the agent startup script
(scriptFile), and the path to the list of table OIDs corresponding to this
instance are passed as arguments.

8) When deleting auxiliary files in DoCleanup, the files with the list of table
OIDs and the startup script files (scriptFile) are deleted.

9) For --single-data-file and --copy-queue-size modes, the current behavior
remains unchanged.

For more convenient management, all auxiliary files have been renamed. The
process pid is placed before the suffix. The word pipe has been removed from
the names of error and skip files.

New tests have been added and old ones have been adapted.

---
It is easier to view the changes with the "Hide whitespace" option enabled.

### CODE REVIEW

File: restore/data.go [hunk 1]
```diff
@@ -346,7 +361,9 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
 func CreateInitialSegmentPipes(oidList []string, c *cluster.Cluster, connectionPool *dbconn.DBConn, fpInfo filepath.FilePathInfo) int {
 	// Create min(connections, tables) segment pipes on each host
 	var maxPipes int
-	if connectionPool.NumConns < len(oidList) {
+	if !backupConfig.SingleDataFile {
```
Comment (silent-observer) [inline restore/data.go:364]: Maybe add a comment here to emphasize this negation (that we are creating a single pipe only if we are *not* in SingleDataFile mode), this condition looks too similar to maxHelpers above

Comment (RekGRpth) [reply restore/data.go:364]: added

**Review by silent-observer — APPROVED** [review]

File: restore/data.go [hunk 2]
```diff
@@ -215,34 +215,49 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
...
+			maxHelpers = totalTables
+		}
+
+		for helperIdx := 0; helperIdx < maxHelpers; helperIdx++ {
+			// During a larger-to-smaller restore, we need to do multiple passes of
+			// data loading so we assign the batches here.
+			oidList := make([]string, 0)
+			for entryIdx, entry := range dataEntries {
+				if (entryIdx % maxHelpers) != helperIdx {
+					continue
```
Comment (whitehawk) [inline restore/data.go:233]: I think we are wasting here too many operations when filtering out entries. On a real world backup that could be millions of skipped entries for each helperIdx, which we could have avoided. What about the change below?
```diff
diff --git a/restore/data.go b/restore/data.go
index 65e3e3e2..73c5a570 100644
--- a/restore/data.go
+++ b/restore/data.go
@@ -228,18 +228,14 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
                        // During a larger-to-smaller restore, we need to do multiple passes of
                        // data loading so we assign the batches here.
                        oidList := make([]string, 0)
-                       for entryIdx, entry := range dataEntries {
-                               if (entryIdx % maxHelpers) != helperIdx {
-                                       continue
-                               }
-
-                               if entry.IsReplicated {
-                                       oidList = append(oidList, fmt.Sprintf("%d,0", entry.Oid))
+                       for entryIdx := helperIdx; entryIdx < len(dataEntries); entryIdx += maxHelpers {
+                               if dataEntries[entryIdx].IsReplicated {
+                                       oidList = append(oidList, fmt.Sprintf("%d,0", dataEntries[entryIdx].Oid))
                                        continue
                                }
 
                                for b := 0; b < batches; b++ {
-                                       oidList = append(oidList, fmt.Sprintf("%d,%d", entry.Oid, b))
+                                       oidList = append(oidList, fmt.Sprintf("%d,%d", dataEntries[entryIdx].Oid, b))
                                }
                        }

```

Comment (RekGRpth) [reply restore/data.go]: applied

**Review by whitehawk — APPROVED** [review]

**Comment by silent-observer** [comment]
The new test has failed, please look into this

**Comment by RekGRpth** [comment]
> The new test has failed, please look into this

This is not a new test, but an [old](https://github.com/arenadata/gpbackup/commit/55a7aa82c967f45562b36b268dfe26450f7c3986) one. But it became flaky (other runs were successful).

File: filepath/filepath.go [hunk 3]
```diff
@@ -103,13 +103,20 @@ func (backupFPInfo *FilePathInfo) replaceCopyFormatStringsInPath(templateFilePat
...
-	templateFilePath := backupFPInfo.GetSegmentPipePathForCopyCommand()
+func (backupFPInfo *FilePathInfo) GetSegmentPipeFilePath(contentID int, helperIdx ...int) string {
+	templateFilePath := backupFPInfo.GetSegmentPipePathForCopyCommand(helperIdx...)
 	return backupFPInfo.replaceCopyFormatStringsInPath(templateFilePath, contentID)
 }
 
-func (backupFPInfo *FilePathInfo) GetSegmentPipePathForCopyCommand() string {
-	return fmt.Sprintf("<SEG_DATA_DIR>/gpbackup_<SEGID>_%s_pipe_%d", backupFPInfo.Timestamp, backupFPInfo.PID)
+func FormatSuffix(suffix string, helperIdx ...int) string {
+	if len(helperIdx) == 1 {
```
[the remark this answers is not in the document]
Comment (RekGRpth) [reply filepath/filepath.go]: added

File: helper/helper.go [hunk 4]
```diff
@@ -78,7 +78,7 @@ func DoHelper() {
 	}
 	if err != nil {
 		// error logging handled in doBackupAgent and doRestoreAgent
-		errFile := fmt.Sprintf("%s_error", *pipeFile)
+		errFile := strings.Replace(*pipeFile, "pipe", "error", -1)
```
[the remark this answers is not in the document]
Comment (RekGRpth) [reply helper/helper.go]: done

File: filepath/filepath.go [hunk 5]
```diff
@@ -103,13 +103,20 @@ func (backupFPInfo *FilePathInfo) replaceCopyFormatStringsInPath(templateFilePat
...
+		suffix += fmt.Sprintf("%d", helperIdx[0])
+	}
+	return suffix
+}
+
+func (backupFPInfo *FilePathInfo) GetSegmentPipePathForCopyCommand(helperIdx ...int) string {
+	return fmt.Sprintf("<SEG_DATA_DIR>/gpbackup_<SEGID>_%s_%d_%s", backupFPInfo.Timestamp, backupFPInfo.PID, FormatSuffix("pipe", helperIdx...))
```
[the remark this answers is not in the document]
Comment (RekGRpth) [reply filepath/filepath.go:125]: may be, but we should also replace it in replaceCopyFormatStringsInPath function too, which is out of current patch scoupe

**Comment by RekGRpth** [comment]
> Shouldn't this [error file name](https://github.com/arenadata/gpbackup/blob/04d77da559ca7506f6481a7489a3df4af65ea842/utils/plugin.go#L417C4-L417C82) be updated as well?

yes, added

**Review by silent-observer — APPROVED** [review]

**Review by whitehawk — APPROVED** [review]

**Pull request merged**